homeassistant/components/mqtt/client.py

```diff
@@ -517,13 +622,18 @@ async def async_subscribe(
         # Only subscribe if currently connected.
         if self.connected:
             self._last_subscribe = time.time()
-            await self._async_perform_subscriptions(((topic, qos),))
+            await self._async_queue_subscriptions(((topic, qos),))
 
         @callback
         def async_remove() -> None:
             """Remove subscription."""
             self._async_untrack_subscription(subscription)
             self._matching_subscriptions.cache_clear()
+            max_qos: int | None = self._max_qos.get(topic)
+            new_max: int | None = (
+                self._max_qos_from_subscriptions(topic) if max_qos is not None else None
+            )
+            self._max_qos[topic] = new_max or 0
```

Comment:
If new_max is None, then no subscriptions had any QoS - which would mean there are no remaining subscriptions - we should probably just del self._max_qos[topic] then.

So should be:
```
if new_max is not None:
   self._max_qos[topic] = new_max
else:
   del self._max_qos[topic]
```